File: frontend/src/shared/api/contracts/notifications/messaging.ts
import type { CourseScheduleEntry } from '@/shared/api/contracts/academics/course'
import type {
  IsoDateString,
  IsoDateTimeString,
  UuidString,
} from '@/shared/api/contracts/common/scalarTypes'
import type {
  EmailDeliveryStatus,
  EmailTemplateKey,
} from '@/shared/api/contracts/notifications/emailDelivery'

/**
 * Leaving `department_id` empty targets the whole association, which the backend authorizes with
 * the organization-wide send permission rather than the department one.
 *
 * department_id 为空表示面向全社团，后端据此要求组织级发送权限而非部门级权限。
 */
export interface BroadcastRequest {
  subject: string
  body: string
  department_id?: UuidString | null
  include_alumni: boolean
  preview_only: boolean
}

export interface BroadcastResult {
  recipient_count: number
  sent_count: number
  skipped_count: number
  failed_count: number
  preview_recipients: string[]
}

export interface DailyDigestWeather {
  condition_en: string
  condition_zh: string
  temperature_high_celsius?: number | null
  temperature_low_celsius?: number | null
  precipitation_probability_percent?: number | null
  sunrise?: string | null
  sunset?: string | null
}

export interface DailyDigestActivity {
  title: string
  department_name_zh?: string | null
  starts_at: IsoDateTimeString
  location?: string | null
}

export interface DailyDigestContent {
  digest_date: IsoDateString
  weekday_en: string
  weekday_zh: string
  recipient_name: string
  weather?: DailyDigestWeather | null
  courses: CourseScheduleEntry[]
  activities: DailyDigestActivity[]
  announcements: string[]
}

export interface EmailDeliveryRecord {
  id: UuidString
  recipient_email: string
  template_key: EmailTemplateKey
  subject: string
  status: EmailDeliveryStatus
  failure_reason?: string | null
  sent_at?: IsoDateTimeString | null
  created_at: IsoDateTimeString
}

File: backend/app/domain/notifications/schemas/messaging.py
"""Outbound messaging payloads: manual broadcasts and the daily digest."""

from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.domain.academics.schemas.course import CourseScheduleEntry
from app.domain.notifications.models.email_delivery_log import (
    EmailDeliveryStatus,
    EmailTemplateKey,
)


class BroadcastRequest(BaseModel):
    """A message the presidium or a department lead sends to a roster.

    Leaving ``department_id`` empty targets the whole association and therefore demands the
    organization-wide send permission rather than the department one.

    department_id 为空表示面向全社团，因而要求组织级发送权限而非部门级权限。
    """

    subject: str = Field(min_length=1, max_length=240)
    body: str = Field(min_length=1, max_length=20000)
    department_id: UUID | None = None
    include_alumni: bool = False
    # Preview resolves the recipient list and returns it without sending anything.
    # 预览模式仅解析收件人列表并返回，不实际发送。
    preview_only: bool = False


class BroadcastResult(BaseModel):
    recipient_count: int
    sent_count: int
    skipped_count: int
    failed_count: int
    preview_recipients: list[str] = Field(default_factory=list)


class DailyDigestWeather(BaseModel):
    """Forecast block of the digest, expressed in the units the campus actually uses."""

    condition_en: str
    condition_zh: str
    temperature_high_celsius: float | None = None
    temperature_low_celsius: float | None = None
    precipitation_probability_percent: int | None = None
    sunrise: str | None = None
    sunset: str | None = None


class DailyDigestActivity(BaseModel):
    title: str
    department_name_zh: str | None = None
    starts_at: datetime
    location: str | None = None


class DailyDigestContent(BaseModel):
    """Everything assembled for one member on one day."""

    model_config = ConfigDict(from_attributes=True)

    digest_date: date
    weekday_en: str
    weekday_zh: str
    recipient_name: str
    weather: DailyDigestWeather | None = None
    courses: list[CourseScheduleEntry] = Field(default_factory=list)
    activities: list[DailyDigestActivity] = Field(default_factory=list)
    announcements: list[str] = Field(default_factory=list)


class EmailDeliveryRecord(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    recipient_email: str
    template_key: EmailTemplateKey
    subject: str
    status: EmailDeliveryStatus
    failure_reason: str | None = None
    sent_at: datetime | None = None
    created_at: datetime
